### DISCUSSION TITLE
How to specify host binary for ` dotnet publish -c release -r osx-x64 -p:UseAppHost=true`

### DISCUSSION THREAD

**Discussion by lanyusan** [dotnet/runtime#79948] [Q&A]
I am trying to publish an Mac app for code signing.

I have tried with `PublishReadyToRun `, which caused the error as in:
 
https://github.com/dotnet/runtime/discussions/79947

Now I am trying to use publish single file with this very simple setup:
```
<PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Release|AnyCPU' ">
    <PublishSingleFile>true</PublishSingleFile>
</PropertyGroup>
```

and command `dotnet publish -c release -r osx-x64`,

which caused the error below:

```
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(104,5): error NETSDK1098: Applications published to a single-file are required to use the application host. You must either set PublishSingleFile to false or set UseAppHost to true.
```

So I ran this command ` dotnet publish -c release -r osx-x64 -p:UseAppHost=true`, which caused this error:

```
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018: The "GenerateBundle" task failed unexpectedly. 
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018: System.ArgumentException: Invalid input specification: Must specify the host binary.
```

How do I set host binary? I couldn't find any documentation by Googling or reading dotnet documentation.

Help much appreciated. I have been stuck here for a couple of days.

**Comment by lanyusan** [comment]
Referencing this issue:

https://github.com/dotnet/runtime/issues/72559

run with `dotnet publish -c release -r osx-x64 -p:UseAppHost=true -p:PublishAot=false` still the same error.

**Reply by am11** [reply]
Could you share a repro? Does it reproduce on another machine?

I'd try publishing a simple app to rule out any installation issue. e.g. this osx-x64 publish works on osx-arm64 system just fine (using .NET 7.0):

```sh
$ dotnet new console -n consoleapp1
$ dotnet publish consoleapp1 -c Release -r osx-x64 --self-contained -p:PublishSingleFile=true -o publish-dir

$ ls publish-dir
consoleapp1	consoleapp1.pdb

$ file publish-dir/consoleapp1
publish-dir/consoleapp1: Mach-O 64-bit executable x86_64
```

**Comment by lanyusan** [comment]
@am11 

I have created a reproducer:

https://github.com/lanyusan/dotnet-publish-bug.git

Clone it and run: `dotnet publish -c release -r osx-x64 --self-contained true -p:UseAppHost=true`

I use a MacBook Air M1 chip, with Ventura 13.1

**Reply by lanyusan** [reply]
I tested it on Ubuntun 22.04 as well, with following command: 

dotnet publish -c release -p:PublishSingleFile=true -p:PublishReadyToRun=true -r linux-x64

And it worked. It seems to be a net6.0-macos specific problem.

**Reply by am11** [reply]
Yup, it seems like issue with OS-specific `TargetFramework`.

Simple[r] repro:

```sh
$ dotnet new console -n consoleapp1
$ sed -i '' 's|</TargetFramework>|-macos</TargetFramework>|' consoleapp1/consoleapp1.csproj
$ dotnet publish consoleapp1 -c Release -r osx-x64 \
    --self-contained -p:PublishSingleFile=true -o publish-dir -p:ApplicationId=com.abc.def -p:UseAppHost=true

...
  Optimizing assemblies for size may change the behavior of the app. Be sure to test after publishing. See: https://aka.ms/dotnet-illink
  Optimizing assemblies for size. This process might take a while.
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018: The "GenerateBundle" task failed unexpectedly. [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018: System.ArgumentException: Invalid input specification: Must specify the host binary [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018:    at Microsoft.NET.HostModel.Bundle.Bundler.GenerateBundle(IReadOnlyList`1 fileSpecs) [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018:    at Microsoft.NET.Build.Tasks.GenerateBundle.ExecuteCore() [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018:    at Microsoft.NET.Build.Tasks.TaskBase.Execute() [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute() [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
/usr/local/share/dotnet/sdk/7.0.101/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Publish.targets(904,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask) [/Users/adeel/projects/consoleapp1/consoleapp1.csproj]
```

cc @vitek-karas, similar call stack as https://github.com/dotnet/runtime/issues/48112. I think we should also make `-p:UseAppHost=true` default for `net#.#-macos` TFM as it is for `net#.#`.

**Reply by lanyusan** [reply]
@am11 
So is this a bug or is there any workaround to generate single file?

Avalonia apps need to be published as single file for signing other wise in sandbox it would crash.

If it is a bug, should I create an issue? Any chance to get a quick fix?

Thanks.

**Reply by lanyusan** [reply]
Which version of dotnet have you tested it on? It looks it doesn't work for net7.0.

**Reply by hez2010** [reply]
> If it is a bug, should I create an issue?

Yes you can create an issue for better tracking and visibility. 

> Any chance to get a quick fix?

I think many of the team are on Christmas/new year vacation recently so the response may be slower.

**Reply by midmo** [reply]
Hi @lanyusan, Hi everyone, I always encounter the same problem when the project targets net8.0-macos or net9.0-macos. I can't publish a single file. I have the same error and I can't find how to set the app host in the dotnet publish command line. Can you please tell me if a solution exists?
